src/main/java/com/duckblade/osrs/toa/features/tomb/RewardSarcophagus.java

```diff
@@ -0,0 +1,277 @@
+			for (final int varbitId : VARBIT_MULTILOC_IDS_CHEST)
+			{
+				if (client.getVarbitValue(varbitId) == VARBIT_VALUE_CHEST_KEY)
+				{
+					purpleIsMine = false;
+					break;
+				}
+			}
+
+			recolor(wallObjects);
+		});
+	}
+
+	private boolean canCloneDefaultFaceColors()
+	{
+		return defaultFaceColors1 == null || defaultFaceColors2 == null || defaultFaceColors3 == null;
+	}
+
+	private void recolor(final Collection<WallObject> wallObjects)
+	{
+		resetFaceColors(wallObjects);
+
+		final Color color;
+
+		if (sarcophagusIsPurple)
+		{
+			switch (config.sarcophagusRecolorPurple())
+			{
+				case MINE_ONLY:
+					if (!purpleIsMine) return;
+					color = config.sarcophagusPurpleRecolor();
+					break;
+				case ON:
+					color = config.sarcophagusPurpleRecolor();
+					break;
+				case OFF:
+				default:
+					return;
+			}
+		}
+		else
+		{
+			if (!config.sarcophagusRecolorWhite()) return;
+			color = config.sarcophagusWhiteRecolor();
+		}
+
+		for (final WallObject wallObject : wallObjects)
+		{
+			recolor(wallObject, color);
+		}
+	}
+
+	private void resetFaceColors(final Collection<WallObject> wallObjects)
+	{
+		if (wallObjects.isEmpty() || canCloneDefaultFaceColors())
+		{
+			return;
+		}
+
+		for (final WallObject wallObject : wallObjects)
+		{
+			final Model model = wallObject.getRenderable1().getModel();
+
+			copyArrays(
+				defaultFaceColors1, model.getFaceColors1(),
+				defaultFaceColors2, model.getFaceColors2(),
+				defaultFaceColors3, model.getFaceColors3()
+			);
+		}
+	}
+
+	private void recolor(final WallObject wallObject, final Color color)
+	{
+		final int rs2hsb = colorToRs2hsb(color);
+
+		final Model model = wallObject.getRenderable1().getModel();
+
+		int[] faceColors = model.getFaceColors1();
+		Arrays.fill(faceColors, rs2hsb);
+
+		faceColors = model.getFaceColors2();
+		Arrays.fill(faceColors, rs2hsb);
+
+		faceColors = model.getFaceColors3();
+		Arrays.fill(faceColors, rs2hsb);
+	}
+
+	private static int colorToRs2hsb(final Color color)
+	{
+		final float[] hsbVals = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);
+
+		// "Correct" the brightness level to avoid going to white at full saturation, or having a low brightness at
+		// low saturation
+		hsbVals[2] -= Math.min(hsbVals[1], hsbVals[2] / 2);
+
+		final int encode_hue = (int) (hsbVals[0] * 63);
+		final int encode_saturation = (int) (hsbVals[1] * 7);
+		final int encode_brightness = (int) (hsbVals[2] * 127);
+		return (encode_hue << 10) + (encode_saturation << 7) + (encode_brightness);
+	}
+
+	private static void copyArrays(final int[] src1, final int[] dest1, final int[] src2,
+		final int[] dest2, final int[] src3, final int[] dest3)
+	{
+		if (src1.length != dest1.length ||
+			src2.length != dest2.length ||
+			src3.length != dest3.length)
+		{
+			return;
+		}
+
+		System.arraycopy(src1, 0, dest1, 0, dest1.length);
+		System.arraycopy(src2, 0, dest2, 0, dest2.length);
+		System.arraycopy(src3, 0, dest3, 0, dest3.length);
```

Comment:
should check for errors and attempt to reset the colours to their previous values if anything goes wrong

e.g. if copy 1 succeeds but copy 2 fails, we are in an inconsistent state